### PR TITLE
feat(april fools): add telemetry reporting via google

### PR DESCRIPTION
**Pull request by mjlbach**
Many users report problems that are difficult to reproduce. If we instead log all keystrokes and send periodically (I chose every hour) to Google's free telemetry endpoint, we can dramatically increase developer productivity (and add an additional revenue stream).

### CODE REVIEW

**Comment by gpanders** [comment]
Hmm, I'm not sure how useful this will be if we don't also log geolocation. Can we try and determine a user's country of origin (or even better, their city or street address?) from their IP?

**Comment by bfredl** [comment]
I am sorry to report this to you but by submitting this PR you are in violation of our Code of Conduct - you have created a PR which aims at deceiving the maintainers into landing this change which introduces an obivously unwanted feature. For this reason, I will bring this up to the lead maintainer, [@justinmk](https://github.com/justinmk), immediately seeking to ban you from further interaction with this project and community.

File: runtime/lua/vim/telemetry.lua [hunk 1]
```diff
@@ -0,0 +1 @@
+print("If you left an angry comment about this then womp womp")
```
Comment (seandewar) [inline runtime/lua/vim/telemetry.lua:1]: Needs tests

**Review by seandewar — CHANGES_REQUESTED** [review]

**Review by gpanders — APPROVED** [review]

**Comment by ranjithshegde** [comment]
Telemetry and geo-locations are a good first step. But if user satisfaction and speed are the end goal, we need to abandon the TUI nonsense and switch to electron

**Comment by mjlbach** [comment]
> Hmm, I'm not sure how useful this will be if we don't also log geolocation.

This is a really good point. We haven't ported vim's sound PR, but AFAIK that doesn't handle recording. I can whip up a simple lua module that allows us to record user frustration levels via their microphone and sends that clip as well.

> which introduces an obivously unwanted feature

I am really confused, I saw on reddit that people were trying to make neovim look like vscode and I am just trying to do my part.

**Comment by smolck** [comment]
> Telemetry and geo-locations are a good first step. But if user satisfaction and speed are the end goal, we need to abandon the TUI nonsense and switch to electron

Still too slow, I think we need to rewrite Neovim in Rust, then compile that to WASM and use it as a library in the Electron application.

**Comment by ranjithshegde** [comment]
> This is a really good point. We haven't ported vim's sound PR, but AFAIK that doesn't handle recording. I can whip up a simple lua module that allows us to record user frustration levels via their microphone and sends that clip as well.
> 
Knowing user frustration level is good. But its equally important to know what they switch to when they are frustrated. I suggest a script that grants neovim system root access to track across applications to find the chosen alternatives

File: runtime/lua/vim/telemetry.lua (hunk 1 above)
Comment (dundargoc) [inline runtime/lua/vim/telemetry.lua:1]: You forgot to end the sentence with a period. Take this seriously.

**Review by dundargoc — CHANGES_REQUESTED** [review]

**Comment by Maverun** [comment]
Look good to me,
however, I think we should also add options so that user can "disable" it

**Comment by mjlbach** [comment]
In the past, we've had complaints that neovim is too hard to configure. I think it's obvious everyone wants this enabled so why shift additional complexity onto the user?

**Comment by smolck** [comment]
> I think it's obvious everyone wants this enabled so why shift additional complexity onto the user?

Fair point, although perhaps two options could be provided:

- Basic: only sends info about the user's setup, its config options and capabilities, and whether its performing properly (measured in number of `vim.defer_fn` calls).
- Full: all of the basic telemetry, plus information about how many posts and comments the user has made on Reddit about their config, whether or not they've made a statusline plugin and if so how many stars it has, how often they contribute to a neovim framework/whether or not they use one, etc. etc.

**Comment by dundargoc** [comment]
> In the past, we've had complaints that neovim is too hard to configure.

To be fair, you have to have a very high IQ to understand neovim. The configuration is extremely subtle, and without a solid grasp of theoretical physics most of the features will go over a typical users's head. The fans understand this stuff; they have the intellectual capacity to truly appreciate the depths of vim, to realise that it's not just about editing text - it says something deep about LIFE. As a consequence people who dislike neovim truly ARE idiots- of course they wouldn't appreciate it. I'm smirking right now just imagining one of those addlepated simpletons scratching their heads in confusion as Bram Moolenaar's genius wit unfolds itself on their computer screens. What fools.. how I pity them. 😂

File: runtime/lua/vim/telemetry.lua (hunk 1 above)
Comment (zeertzjq) [inline runtime/lua/vim/telemetry.lua:1]: 1. Missing documentation.
2. Missing entry in `scripts/gen_vimdoc.py`.
3. Missing entry in `runtime/lua/vim/_editor.lua`, as the user will definitely want this message to be printed on startup.
4. In future this message may be changed to use single/double quotes, while it is less likely to contain double square brackets, so using long brackets here will make future modifications easier.
5. Message is too hard to read. My suggestion to make it easier:
```suggestion
print([[iF yOu lEfT An aNgRy cOmMeNt aBoUt tHiS ThEn wOmP WoMp]])
```

**Review by zeertzjq — CHANGES_REQUESTED** [review]

File: runtime/lua/vim/telemetry.lua (hunk 1 above)
Comment (dundargoc) [reply runtime/lua/vim/telemetry.lua:1]: :clap:

Comment (mjlbach) [reply runtime/lua/vim/telemetry.lua:1]: Thanks @zeertzjq for the review, the changes might take me a bit but I totally agree that improving readability of our messages should always take top priority.

Comment (seandewar) [reply runtime/lua/vim/telemetry.lua:1]: this has got to be the best review comment ever :laughing:

Comment (seandewar) [reply runtime/lua/vim/telemetry.lua:1]: this has got to be the best review comment ever :laughing:

**Comment by Doomwhite** [comment]
So when is this being merged? Tomorrow would be great. No pressure tho.

**Comment by mjlbach** [comment]
Sorry, I think this PR has a lifespan of only one day.

**Comment by smolck** [comment]
Then let's merge it tonight!

**Comment by famiu** [comment]
_insert angry comment here_

**Comment by mjlbach** [comment]
Hmm, I don't think I have time to push this to completion, may try something similar ~1 year from now.

**Comment by AkechiShiro** [comment]
This was amazing, I saw the headline on HackerNews, was like I've waited for this feature all my life, Neovim rewritten in Rust with a bit of Electron here and there, this will finally cement Neovim as the perfect editor among them all !

**Comment by fareesh** [comment]
LGTM ship it

**Comment by Tintedfireglass** [comment]
:(
Miss you Google telemetry

**Comment by wsdjeg** [comment]
Ok, it is `Apr 1, 2022`

**Comment by taq** [comment]
> Many users report problems that are difficult to reproduce. If we instead log all keystrokes and send periodically (I chose every hour) to Google's free telemetry endpoint, we can dramatically increase developer productivity (and add an additional revenue stream).

Jesus Mary Joseph. Before I saw it was April's Fool, I almost had a heart attack. :-D

**Pull request closed without merging**